samples/nrf9160/https_client/src/main.c

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <string.h>
+#include <zephyr.h>
+#include <stdlib.h>
+#include <net/socket.h>
+#include <net/bsdlib.h>
+#include <net/tls_credentials.h>
+#include <lte_lc.h>
+#include <at_cmd.h>
+#include <at_notif.h>
+#include <modem_key_mgmt.h>
+
+#define HTTPS_PORT 443
+
+#define HTTP_HEAD                                                              \
+	"HEAD / HTTP/1.1\r\n"                                                  \
+	"Host: www.google.com:443\r\n"                                         \
+	"Connection: close\r\n\r\n"
+
+#define RECV_BUF_SIZE 2048
+#define TLS_SEC_TAG 42
+
+static const char send_buf[] = HTTP_HEAD;
+static char recv_buf[RECV_BUF_SIZE];
+
+/* Certificate for `google.com` */
+static const char cert[] = {
+	#include "../cert/GlobalSign-Root-CA-R2"
+};
+
+BUILD_ASSERT_MSG(sizeof(cert) < KB(4), "Certificate too large");
+
+
+/* Initialize AT communications */
+int at_comms_init(void)
+{
+	int err;
+
+	err = at_cmd_init();
+	if (err) {
+		printk("Failed to initialize AT commands, err %d\n", err);
+		return err;
+	}
+
+	err = at_notif_init();
+	if (err) {
+		printk("Failed to initialize AT notifications, err %d\n", err);
+		return err;
+	}
+
+	return 0;
+}
+
+/* Provision certificate to modem */
+int cert_provision(void)
+{
+	int err;
+	bool exists;
+	u8_t unused;
+
+	printk("Checking certificate..\n");
+
+	err = modem_key_mgmt_exists(TLS_SEC_TAG,
+				    MODEM_KEY_MGMT_CRED_TYPE_CA_CHAIN,
+				    &exists, &unused);
+	if (err) {
+		printk("Failed to check for certificates err %d\n", err);
+		return err;
+	}
+
+	if (exists) {
+		/* For the sake of simplicity we delete is provisioned
+		 * with our security tag and reprovision our certificate.
+		 */
+		err = modem_key_mgmt_delete(TLS_SEC_TAG,
+					    MODEM_KEY_MGMT_CRED_TYPE_CA_CHAIN);
+		if (err) {
+			printk("Failed to delete existing certificate, err %d\n",
+			       err);
+		}
+	}
+
+	printk("Provisioning certificate\n");
+
+	/*  Provision certificate to the modem */
+	err = modem_key_mgmt_write(TLS_SEC_TAG,
+				   MODEM_KEY_MGMT_CRED_TYPE_CA_CHAIN,
+				   cert, sizeof(cert) - 1);
+	if (err) {
+		printk("Failed to provision certificate, err %d\n", err);
+		return err;
+	}
+
+	return 0;
+}
+
+/* Setup TLS options on a given socket */
+int tls_setup(int fd)
+{
+	int err;
+	int verify;
+
+	/* Security tag that we have provisioned the certificate with */
+	const sec_tag_t tls_sec_tag[] = {
+		TLS_SEC_TAG,
+	};
+
+	/* Set up TLS peer verification */
+	enum {
+		NONE = 0,
```

Comment:
IMO we should specify these in the upstream `socket.h`. I'll add this to my nit-list.